include/initialGuess.hpp

```diff
@@ -49,120 +48,176 @@ class initialGuessStrategy_t {
 
  public:
   initialGuessStrategy_t(dlong _N, platform_t& _platform, settings_t& _settings, comm_t _comm):
-    platform(_platform), settings(_settings), comm(_comm), Ntotal(_N) {}
+                         platform(_platform), settings(_settings), comm(_comm), Ntotal(_N) {}
+
+  virtual void FormInitialGuess(deviceMemory<float>& o_x, deviceMemory<float>& o_rhs)  { printf("AROEOHAER\n"); }
+  virtual void Update(operator_t& linearOperator, deviceMemory<float>& o_x, deviceMemory<float>& o_rhs) { printf("AROEOHAER\n"); }
 
-  virtual void FormInitialGuess(deviceMemory<dfloat>& o_x, deviceMemory<dfloat>& o_rhs) = 0;
-  virtual void Update(operator_t& linearOperator, deviceMemory<dfloat>& o_x, deviceMemory<dfloat>& o_rhs) = 0;
+  virtual void FormInitialGuess(deviceMemory<double>& o_x, deviceMemory<double>& o_rhs) { printf("AROEOHAER\n"); }
+  virtual void Update(operator_t& linearOperator, deviceMemory<double>& o_x, deviceMemory<double>& o_rhs) { printf("AROEOHAER\n"); }
 };
 
 // Default initial guess strategy:  use whatever the last solution was (starting at the zero vector)
+template <typename T>
 class Last : public initialGuessStrategy_t {
 private:
-  deviceMemory<dfloat> o_xLast;
+  deviceMemory<T> o_xLast;
 
 public:
   Last(dlong _N, platform_t& _platform, settings_t& _settings, comm_t _comm);
 
-  void FormInitialGuess(deviceMemory<dfloat>& o_x, deviceMemory<dfloat>& o_rhs);
-  void Update(operator_t &linearOperator, deviceMemory<dfloat>& o_x, deviceMemory<dfloat>& o_rhs);
+  Last<T>& operator=(const Last<T> &m)=default;
+
+  void FormInitialGuess(deviceMemory<T>& o_x, deviceMemory<T>& o_rhs);
+  void Update(operator_t &linearOperator, deviceMemory<T>& o_x, deviceMemory<T>& o_rhs);
 };
 
+template class Last<double>;
+template class Last<float>;
```

Comment:
Instantiating the class templates here will have the compiler instantiate it in every translation unit that includes this header. We want to `extern template class ...` here, and have these lines which explicitly instantiate in the .cpp files.